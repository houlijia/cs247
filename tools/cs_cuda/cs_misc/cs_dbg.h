#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
void dbg_pdata_i( char *s, int *dp, int size ) ;
void dbg_pdata_c( char *s, char *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( char *s, int *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
void dbg_p_d_data_i_mn_skip ( char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_i_mn_v2 ( char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

// tsting ...
void dbg_p_d_data_ll ( char *s, long long *dp, int size ) ;

#endif 
