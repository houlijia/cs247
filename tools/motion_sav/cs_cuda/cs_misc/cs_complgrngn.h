#ifndef __CS_COMPLGRNGN_H__
#define __CS_COMPLGRNGN_H__

struct sqrerr {
	float A ;		//||(Ax-b)||^2
	float D ;		//||(Dx-w)||^2
} ;

struct xerr {
	float *d_A ;	// 4752x1  (Ax-b)
	int A_size ;

	float *d_D ;	// 75072x1 (Dx-w)
	int D_size ;

	float J ;
} ;

struct lmderr {
	float A ;	// lambda_A'(Ax-b) == lambda_A' * xerr.d_A
	float D ;   // lambda_D'(Dx-w) == lambda_D' * xerr.d_D
} ;

struct lambda {
	float *d_A ;	// 4752x1
	int A_size ;

	float *d_D ;	// 75072x1
	int D_size ;
} ;

// protos

float h_do_compLgrng ( struct beta *betap, struct lmderr *lmderrp, struct sqrerr *sqrerrp,
	struct xerr *xerrp ) ;

int h_do_sqrerr ( struct sqrerr *sqrerrp, struct xerr *xerrp, float *d_p, int tbl_size ) ;
int h_do_lmderr ( struct lmderr *lmderrp, struct lambda *lambdap, struct xerr *xerrp,
	float *d_p, int tbl_size ) ;

#endif 
