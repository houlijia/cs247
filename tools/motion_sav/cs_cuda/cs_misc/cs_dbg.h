#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
int dbg_copy_d_data ( char *dtp, char *dfp, int size ) ;
void dbg_pdata_i( const char *s, const int *dp, int size ) ;
void dbg_pdata_f( const char *s, const float *dp, int size ) ;
void dbg_pdata_c( const char *s, const char *dp, int size ) ;
void dbg_pdata_d( const char *s, const float *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( const char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( const char *s, int *dp, int size ) ;
void dbg_p_d_data_f ( const char *s, float *dp, int size ) ;
void dbg_p_d_data_d ( const char *s, float *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
float *dbg_d_malloc_f ( int size ) ;

void dbg_p_d_data_f_mn_skip ( const char *s, float *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;
void dbg_p_d_data_i_mn_skip ( const char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint, int printrow ) ;
void dbg_p_d_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( const char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

// print the motion detection T/V/H/Va format
void dbg_p_data_md_f_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_f_mn_v2 ( const char *s, float *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_d_data_i_mn_v2 ( const char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( const char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

void dbg_p_d_data_i_cube ( const char *s, int *devp, int vx, int hy, int tz ) ;
void dbg_p_d_data_i_cube ( const char *s, float *devp, int vx, int hy, int tz ) ;
void dbg_p_data_i_cube ( const char *s, float *dp, int vx, int hy, int tz ) ;

void dbg_pr_first_last ( char *s, float *d_p, int len, int pr_size ) ;
void dbg_pr_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, char *h_p, int len, int pr_size ) ;

// tsting ...
void dbg_p_d_data_ll ( const char *s, long long *dp, int size ) ;

// for random
int dbg_ck_unique ( char *s, int *dp, int size ) ;
int dbg_perm_ck ( int callid, int callid2, int zero_in_zero, int *d_bp, int size ) ;

// TVH
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print, int do_skip ) ;
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print,
	int do_skip, int start ) ;

#endif 
