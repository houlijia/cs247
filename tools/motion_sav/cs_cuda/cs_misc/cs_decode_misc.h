#ifndef __CS_DECODE_MISC_H__
#define __CS_DECODE_MISC_H__

struct grad_x_const {
	float *d_A ;	 // 38016 == 72 * 88 * 2 * 3
	float *d_D ;	 // 38016 == 72 * 88 * 2 * 3
	float *d_sum ; 	 // 38016 == 72 * 88 * 2 * 3
} ;

struct Pgrad_x {
	float *d_A ; // 4572 ... but need v*h*t*c*2
	int A_size ;
	float *d_D ; // 75072 in .m ... in .cu it is v*h*t*c*2
	int D_size ;
} ;

struct vhtc {
	int v ;
	int h ;
	int t ;
	int c ;

	int size ; // v * h * t * c
	int size_mod2 ;
	int size_x2 ; 	

	int A_size ; // 4752
	int D_size ; // 75072
} ;

struct final {
	float scldA ;
	float A ;
	float D ;
} ;

struct beta {
	float A ;
	float D ;
	// float final ;
	struct final final ;
	float scldA ;
} ;

struct LRperms {	
	RndC_uint32 *d_Lperm ; // 65536, max_log2( 72x88x2x3 )
	RndC_uint32 *d_Rperm ; // 65536, max_log2( 72x88x2x3 )
} ;

struct cjerr {	
	float *d_A ; // 38016 ... 72x88x2 x3
	float *d_D ; // 38016 ... 72x88x2 x3
} ;

#endif 
