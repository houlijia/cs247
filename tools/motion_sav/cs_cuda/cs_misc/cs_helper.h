#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void set_device_mem_i ( int *d_a, int n, int val ) ;
void clear_device_mem_i ( int *d_a, int n ) ;

void set_device_mem ( float *d_a, int n, float val ) ;
void clear_device_mem ( float *d_a, int n )	 ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;

void h_expand_c_to_i ( char *din, int *dout, int n ) ;	
void h_expand_c_to_i ( unsigned char *din, int *dout, int n ) ;	
template<typename F, typename T> void h_cast_T_to_T ( F *din, T *dout, int n ) ;

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;
int h_set_cube_config( struct cube *d_a, struct cube *hp ) ;
void h_p_d_cube_config ( const char *, struct cube *dp ) ;
void h_p_h_cube_config ( const char *, struct cube *dp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int put_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes
int get_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes

void h_do_int_to_float ( int *intp, float *fp, int tbl_size ) ;

#endif 
