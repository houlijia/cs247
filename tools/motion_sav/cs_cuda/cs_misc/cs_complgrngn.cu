#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <math.h>
#include "cs_dbg.h"
#include "cs_cuda.h"
#include "cs_helper.h"
#include "cs_matrix.h"

#include "RndCState.h"
#include "RndC_ifc.h"

#include "cs_decode_misc.h"
#include "cs_complgrngn.h"

// #define CUDA_DBG
#define CUDA_DBG1

float
h_do_compLgrng ( struct beta *betap, struct lmderr *lmderrp, struct sqrerr *sqrerrp,
	struct xerr *xerrp )
{
	float lgr ;

	lgr = xerrp->J + lmderrp->D + betap->D * 0.5 * sqrerrp->D +
		lmderrp->A + betap->scldA * 0.5 * sqrerrp->A ;

	return ( lgr ) ;
}

int
h_do_sqrerr ( struct sqrerr *sqrerrp, struct xerr *xerrp, float *d_p, int tbl_size )
{
	if (( tbl_size < xerrp->A_size ) || ( tbl_size < xerrp->D_size ))
	{
		printf("%s: error tbl %d A %d D %d \n", __func__, tbl_size, xerrp->A_size, xerrp->D_size ) ;
		return ( 0 ) ;
	}

	sqrerrp->A = h_do_dot ( xerrp->d_A, xerrp->d_A, d_p, xerrp->A_size ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f ( "h_do_sqrerr.A", d_p, xerrp->A_size );
#endif 

	sqrerrp->D = h_do_dot ( xerrp->d_D, xerrp->d_D, d_p, xerrp->D_size ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f ( "h_do_sqrerr.D", d_p, xerrp->D_size );
#endif 

	return ( 1 ) ;
}

int
h_do_lmderr ( struct lmderr *lmderrp, struct lambda *lambdap, struct xerr *xerrp,
	float *d_p, int tbl_size )
{
	if (( tbl_size < xerrp->A_size ) || ( tbl_size < xerrp->D_size ) ||
		( lambdap->A_size != xerrp->A_size ) || ( lambdap->D_size != xerrp->D_size ))
	{
		printf("%s: error tbl %d xerr A %d D %d lambda A %d D %d\n", __func__, 
			tbl_size, xerrp->A_size, xerrp->D_size, lambdap->A_size, lambdap->D_size ) ;
		return ( 0 ) ;
	}

	lmderrp->A = h_do_dot ( lambdap->d_A, xerrp->d_A, d_p, xerrp->A_size ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f ( "h_do_lmderr.A", d_p, xerrp->A_size );
#endif 

	lmderrp->D = h_do_dot ( lambdap->d_D, xerrp->d_D, d_p, xerrp->D_size ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f ( "h_do_lmderr.D", d_p, xerrp->D_size );
#endif 

	return ( 1 ) ;
}
