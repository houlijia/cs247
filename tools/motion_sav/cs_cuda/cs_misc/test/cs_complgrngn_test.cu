#include <iostream>
using namespace std;

#include <thrust/reduce.h>
#include <thrust/sequence.h>
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <math.h>

#include "RndCState.h"
#include "RndC_ifc.h"

#include "cs_dbg.h"
#include "cs_helper.h"
#include "cs_decode_misc.h"
#include "cs_complgrngn.h"

#define CUDA_DBG
#define BUF_SIZE_INT		10

float xerr_A[] = { 2.1, 3.1, 4.1, 5.2 } ;
float xerr_D[] = { 2.2, 3.2, 4.2, 5.2, 6.2, 7.2, 8.2, 9.2 } ;

float lambda_A[] = { 0.1, 2.1, 3.1, 4.1 } ;
float lambda_D[] = { 1.0, 2.0, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8 } ;

struct beta beta ;
struct lmderr lmderr ;
struct sqrerr sqrerr ;
struct xerr xerr ;
struct lambda lambda ;

int
main( int ac, char *av[] )
{
	int k ;
	float lgr, *dp, *xA, *xD, *lA, *lD ;

	setbuf( stdout, NULL ) ;
	setbuf( stderr, NULL ) ;

	dbg_init ( 1024 * 1024 ) ;



	if (( k = cudaMalloc( &xA, sizeof ( float ) * ( 24 +  100 ) )) != cudaSuccess )
	{
		printf("%s: cube alloc failed %d \n", __func__, k ) ;
		exit ( 0 ) ;
	}

	xD=xA + 4 ;
	lA =xD + 8 ;
    lD = lA + 4 ;	
	dp = lD + 8 ;

	put_d_data_f ( xA, xerr_A, sizeof ( float ) *4 ) ; 
	put_d_data_f ( xD, xerr_D, sizeof ( float ) *8 ) ; 

	put_d_data_f ( lA, lambda_A, sizeof ( float ) *4 ) ; 
	put_d_data_f ( lD, lambda_D, sizeof ( float ) *8 ) ; 

	beta.A = 1.1 ;
	beta.D = 2.2 ;
	// beta.final = 3.3 ;
	beta.scldA = 4.4 ;

	lmderr.A = 2.2 ;
	lmderr.D = 3.3 ;

	sqrerr.A = 4.4 ;
	sqrerr.D = 5.5 ;

	xerr.d_A = xA ;
	xerr.A_size = 4 ;
	xerr.d_D = xD ;
	xerr.D_size = 8 ;
	xerr.J = 9.9 ;

	lambda.d_A = lA ;
	lambda.A_size = 4 ;
	lambda.d_D = lD ;
	lambda.D_size = 8 ;

#ifdef CUDA_OBS 
	dbg_p_d_data_f ( "xerr.A", xerr.d_A, 4 ); 
	dbg_p_d_data_f ( "xerr.D ", xerr.d_D, 8 ); 
	printf("sqrerr A %f D %f \n", sqrerr.A, sqrerr.D ) ;
	printf("-------------------------------------------------------------- 1\n") ;
#endif 

	h_do_sqrerr( &sqrerr, &xerr, dp, 100 ) ;

#ifdef CUDA_OBS 
	dbg_p_d_data_f ( "xerr.A", xerr.d_A, 4 ); 
	dbg_p_d_data_f ( "xerr.D ", xerr.d_D, 8 ); 
#endif 
	printf("sqrerr A %f == 57.869995 D %f == 301.919983\n", sqrerr.A, sqrerr.D ) ;
#ifdef CUDA_OBS 
	printf("-------------------------------------------------------------- 2\n") ;
#endif 

	// test end

#ifdef CUDA_OBS 
	dbg_p_d_data_f ( "xerr.A", xerr.d_A, 4 ); 
	dbg_p_d_data_f ( "xerr.D ", xerr.d_D, 8 ); 
	dbg_p_d_data_f ( "lambda.A", lambda.d_A, 4 ); 
	dbg_p_d_data_f ( "lambda.D ", lambda.d_D, 8 ); 
	printf("lmderr A %f D %f \n", lmderr.A, lmderr.D ) ;
	printf("-------------------------------------------------------------- 3\n") ;
#endif 

	h_do_lmderr( &lmderr, &lambda, &xerr, dp, 100 ) ;

#ifdef CUDA_OBS 
	dbg_p_d_data_f ( "xerr.A", xerr.d_A, 4 ); 
	dbg_p_d_data_f ( "xerr.D ", xerr.d_D, 8 ); 
	dbg_p_d_data_f ( "lambda.A", lambda.d_A, 4 ); 
	dbg_p_d_data_f ( "lambda.D ", lambda.d_D, 8 ); 
#endif 
	printf("lmderr A %f == 40.749996D %f == 271.059998\n", lmderr.A, lmderr.D ) ;
#ifdef CUDA_OBS 
	printf("-------------------------------------------------------------- 4\n") ;
#endif 

	lgr = h_do_compLgrng( &beta, &lmderr, &sqrerr, &xerr ) ;

	printf("lgr %f == 781.135986\n", lgr ) ; 

	// dbg_p_d_data_f ( "print 4 ", fp1, 4 ); 

	// get_d_data_f ( fp1, dbuf1, 4 * sizeof ( float )) ;
	// dbg_pdata_f ( "print 4 ", dbuf1, 4 ); 
}
