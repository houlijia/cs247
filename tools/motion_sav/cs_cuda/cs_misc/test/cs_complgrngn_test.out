Tue Dec 30 14:43:35 EST 2014
cs_complgrngn_test
---------------------------------------------------
sqrerr A 57.869995 == 57.869995 D 301.919983 == 301.919983
lmderr A 40.749996 == 40.749996D 271.059998 == 271.059998
lgr 781.135986 == 781.135986
